src/main/kotlin/com/wandera/affectedmodulesdetector/graph/GraphFinder.kt

```diff
@@ -0,0 +1,45 @@
+package com.wandera.affectedmodulesdetector.graph
+
+/**
+ * Class searching through graph.
+ *
+ * @property graph Graph to search through.
+ */
+class GraphFinder(
+    private val graph: Graph,
+) {
+
+    /**
+     * Finds all nodes that are dependent on given nodes.
+     *
+     * @param nodes Nodes to find dependent nodes for.
+     * @return Input nodes and nodes dependent on them.
+     */
+    fun findDependentNodes(nodes: Set<Node>): Set<Node> {
+        val dependentNodes = mutableSetOf(*nodes.toTypedArray())
+        findDependentNodesRecursive(nodes, dependentNodes)
+        return dependentNodes
+    }
+
+    private fun findDependentNodesRecursive(
```

Comment:
While it makes sense to choose a recursive implementation here for graph processing, it is generally difficult to fully comprehend and should be as simple as possible.
We should also avoid allocations in individual recursion steps and only allocate in the terminal step. In our case we use just below 30 modules, but it could get very ugly very quickly.

I took a liberty to come up with such solution
```kotlin
private fun findDependentNodesRecursive(
        nodes: Set<String>,
    ): Set<String> = if (nodes.isEmpty()) {
        emptySet()
    } else {
        nodes + findDependentNodesRecursive(
            nodes.map { node: String -> getChildren(node) }
                .flatten()
                .toSet()
        )
    }


    private fun getChildren(node: String): Set<String> {
        return edges.filter { it.first == node }.map { it.second }.toSet()
    }
```
It is up to you, it is your code after all. Just consider the risks of being publicly available for larger projects.

Reply:
I for using some better more clear solution. Just if I'm correct wouldn't this algorithm run infinitely if there is cyclic dependency between modules.

Reply:
True. My suggestion in that case would be to get rid of that cyclic dependency, though. That is what dependency inversion is for. There is none in our project and we should not encourage others to do such things.
I would not complicate the implementation to support bad practices.